### PR TITLE
Reports correct sum with currencies and label

### PR DESCRIPTION
**Pull request by fcaraballo**
| Q             | A
| ------------- | ---
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Fixed tickets | #3715
| License       | MIT
| Doc PR        | no

- Added field exchange_rate to sylius_order in order to save the exchange rate for the time the order was placed
- Updated query for reports in order to count check the exchange rate (o.total * o.exchange_rate)
- Behat for reports feature "Getting the correct sum with exchange rate"
- Table label: if there's a base currency it will show it. For e.g. TOTAL SUM IN GBP
- Created method for getting the channel base currency (getBaseCurrency)
- Behats and PHPSpec updated

### CODE REVIEW

File: src/Sylius/Bundle/CoreBundle/spec/EventListener/OrderCurrencyListenerSpec.php [hunk 1]
```diff
@@ -52,4 +52,4 @@ function it_sets_currency_on_order(
 
         $this->processOrderCurrency($event);
     }
-}
+}
```
Comment (stloyd) [inline src/Sylius/Bundle/CoreBundle/spec/EventListener/OrderCurrencyListenerSpec.php]: Should be reverted.

File: src/Sylius/Component/Core/Repository/OrderRepositoryInterface.php [hunk 2]
```diff
@@ -54,11 +54,12 @@ public function countByCustomerAndPaymentState(CustomerInterface $customer, $sta
...
      * between particular dates
      *
-     * @param array $configuration
+     * @param string $baseCurrency
+     * @param array  $configuration
      *
      * @return array
      */
-    public function revenueBetweenDatesGroupByDate(array $configuration = array());
+    public function revenueBetweenDatesGroupByDate($baseCurrency, array $configuration = array());
```
Comment (stloyd) [inline src/Sylius/Component/Core/Repository/OrderRepositoryInterface.php]: Why new parameter and not just additional configuration field?

File: src/Sylius/Bundle/CoreBundle/Checkout/Step/FinalizeStep.php [hunk 3]
```diff
@@ -63,11 +63,19 @@ protected function renderStep(ProcessContextInterface $context, OrderInterface $
      */
     protected function completeOrder(OrderInterface $order)
     {
+        $currencyProvider = $this->get('sylius.currency_provider');
+        $currencyRepository = $this->get('sylius.repository.currency');
```
Comment (stloyd) [inline src/Sylius/Bundle/CoreBundle/Checkout/Step/FinalizeStep.php]: It's use only if the `if` below is true, so it should be initialized there.

File: features/reports/reports.feature [hunk 4]
```diff
@@ -140,3 +141,44 @@ Feature: Reports
...
+        And the following products exist:
+            | name | price | sku |
+            | Mug  | 10    | 456 |
+            | Book | 22    | 948 |
+        And the following orders exist:
+            | customer                | shipment                 | address                                                           | currency | exchange_rate |
+            | sylius@example.com      | UPS, shipped, DTBHH380HG | Théophile Morel, 17 avenue Jean Portalis, 33000, Bordeaux, France | EUR      | 1.00000       |
+            | linustorvalds@linux.com | DHL, shipped, DTBHH380HH | Linus Torvalds, Väätäjänniementie 59, 00440, Helsinki, Finland    | USD      | 0.5           |
+            | sylius@example.com      | UPS, shipped, DTBHH380HI | Théophile Morel, 17 avenue Jean Portalis, 33000, Bordeaux, France | GBP      | 1.25          |
+        And order #000000001 has following items:
```
Comment (stloyd) [inline features/reports/reports.feature]: IMO it should be `"#000000001"` so it won't be rendered as comment in diff's.

**Comment by peteward** [comment]
Remember to a description of these changes to `UPGRADE.md`

File: features/reports/reports.feature (hunk 4 above)
Comment (fcaraballo) [inline features/reports/reports.feature]: Sylius has the steps on this way, I will leave it for consistency 

Sylius e.g.
And order #000000001 has following items:
          | product | quantity |
          | Mug     | 2        |

**Comment by fcaraballo** [comment]
Updated for review

File: src/Sylius/Bundle/CoreBundle/DataFetcher/SalesTotalDataFetcher.php [hunk 5]
```diff
@@ -28,7 +28,7 @@ class SalesTotalDataFetcher extends TimePeriod
     private $orderRepository;
 
     /**
-     * @param OrderRepositoryInterface $orderRepository
+     * @param OrderRepositoryInterface  $orderRepository
```
Comment (stloyd) [inline src/Sylius/Bundle/CoreBundle/DataFetcher/SalesTotalDataFetcher.php]: Should be reverted.

File: src/Sylius/Bundle/CoreBundle/Checkout/Step/FinalizeStep.php [hunk 6]
```diff
@@ -63,12 +63,19 @@ protected function renderStep(ProcessContextInterface $context, OrderInterface $
...
     {
-        $this->get('session')->set('sylius_order_id', $order->getId());
+        $currencyProvider = $this->get('sylius.currency_provider');
+        
         $this->dispatchCheckoutEvent(SyliusOrderEvents::PRE_CREATE, $order);
         $this->dispatchCheckoutEvent(SyliusCheckoutEvents::FINALIZE_PRE_COMPLETE, $order);
 
         $this->get('sm.factory')->get($order, OrderTransitions::GRAPH)->apply(OrderTransitions::SYLIUS_CREATE, true);
-
+
```
Comment (stloyd) [inline src/Sylius/Bundle/CoreBundle/Checkout/Step/FinalizeStep.php]: Extra indent should be removed.

File: src/Sylius/Component/Core/Repository/OrderRepositoryInterface.php [hunk 7]
```diff
@@ -54,7 +54,7 @@ public function countByCustomerAndPaymentState(CustomerInterface $customer, $sta
      * Gets revenue group by date
      * between particular dates
      *
-     * @param array $configuration
+     * @param array  $configuration
```
Comment (stloyd) [inline src/Sylius/Component/Core/Repository/OrderRepositoryInterface.php]: Should be reverted.

File: src/Sylius/Bundle/ReportBundle/Controller/ReportController.php [hunk 8]
```diff
@@ -70,6 +72,8 @@ public function embedAction(Request $request, $report, array $configuration = ar
         }
 
         $configuration = $request->query->get('configuration', $configuration);
+        $configuration['baseCurrency']= $currencyProvider->getBaseCurrency();
```
Comment (stloyd) [inline src/Sylius/Bundle/ReportBundle/Controller/ReportController.php]: IMO it should be `base_currency`.

Comment (stloyd) [inline src/Sylius/Bundle/ReportBundle/Controller/ReportController.php]: Missing space before `=`.

Comment (fcaraballo) [inline src/Sylius/Bundle/ReportBundle/Controller/ReportController.php]:
```
array:9 [▼
  "start" => DateTime {#1679 ▶}
  "end" => DateTime {#1696 ▶}
  "period" => "month"
  "empty_records" => false
  "baseCurrency" => Currency {#3310 ▶}
  "interval" => "P1M"
  "periodFormat" => "%m"
  "presentationFormat" => "F Y"
  "groupBy" => array:2 [▶]
]
```
seems to be mixed, I will leave it like this, but yeah has to have consistency at some point

File: src/Sylius/Component/Core/Currency/ChannelAwareCurrencyProvider.php [hunk 9]
```diff
@@ -11,29 +11,32 @@
...
      * @var ChannelContextInterface
      */
     protected $channelContext;
 
     /**
      * @param ChannelContextInterface $channelContext
+     * @param EntityRepository $currencyRepository
      */
-    public function __construct(ChannelContextInterface $channelContext)
+    public function __construct(ChannelContextInterface $channelContext, EntityRepository $currencyRepository)
```
Comment (adamelso) [inline src/Sylius/Component/Core/Currency/ChannelAwareCurrencyProvider.php]: This should be `Sylius\Component\Resource\Repository\RepositoryInterface` to match the parent and allow for non-ORM repositories to be injected.

**Comment by fcaraballo** [comment]
Updated for review

**Comment by pjedrzejewski** [comment]
@fcaraballo Could you rebase, please? :) Thanks!

File: src/Sylius/Bundle/CoreBundle/Behat/CoreContext.php [hunk 10]
```diff
@@ -148,7 +148,16 @@ public function thereAreOrders(TableNode $table)
 
             $this->createPayment($order, $paymentMethod);
 
-            $order->setCurrency('EUR');
+            if (isset($data['currency']) && '' !== trim($data['currency'])) {
```
Comment (pjedrzejewski) [inline src/Sylius/Bundle/CoreBundle/Behat/CoreContext.php]: Why you need to check and add trim?

```php
$currency = isset($data['currency']) ? trim($data['currency']) : 'EUR';
$order->setCurrency($currency);
```

Should be enough?

**Comment by pjedrzejewski** [comment]
Thanks Fernando! :+1:

**Pull request merged**